### ISSUE TITLE
No Changes In Learned Structures

### ISSUE THREAD

**Issue by H-arshit** [google-research/morph-net#103]
Am getting all the learned structures same as my initial network , i.e morphnet did not change my network at all. what could be the possible reasons for this.. and how to correct it ?

**Comment by eladeban** [comment]
Hi,

It is a bit hard to say without knowing more about:
1. your model.
2. the regularizer you used (Gamma*, GroupLasso*)
3. the regularization strength, and threshold.

If you are using a model with batch norm, make sure you set `fused` and `scale` to `True`, and then use `GammaFlops` or `GammaModelSize`.
If you don't have batch norm, then you must use `GroupLasso` based regularization. 

Regarding regularization strength, if you don't see any effect, please consider a stronger regularization strength say by a factor of 10x. After you get to a range of regularization strength that has some effect, you should be able to tradeoff resources  and accuracy.

**Comment by H-arshit** [comment]
Hello @eladeban ,
Thankyou for replying 
I am training  **UNet Semantic Segmentation Model** for **COCO dataset** on a small cut of around 6000 training instances with mask of tag  : "Person".
I have implemented the whole model from scratch using **tensorflow.contrib.layers** , having **batch norm** . I am try to reduce the size of the network using **GammaModelSizeRegularizer** . Also as instructed  in the documentation i have set `scale` and `fused` to `True` . 
I have also tested for 2 different `regularization_strength` equals `1e-14` and `1e-10`. 


Code for my above implementation is provided below : 

Defining Different Blocks For Unet 
```python
def down_block(
    input_tensor,
    no_filters,
    kernel_size=(3, 3),
    strides=(1, 1),
    padding="SAME",
    max_pool_window=(2, 2),
    max_pool_stride=(2, 2)):
    
    
    conv = layers.conv2d(
        inputs = input_tensor,
        num_outputs = no_filters,
        normalizer_fn = layers.batch_norm,
        normalizer_params = {"scale" : True , "fused" : True},
        kernel_size = kernel_size,
        stride = strides,
        activation_fn = tf.nn.relu,
        padding = padding)

    conv = layers.conv2d(
        inputs = conv,
        num_outputs = no_filters,
        normalizer_fn = layers.batch_norm,
        normalizer_params = {"scale" : True , "fused" : True},
        kernel_size = kernel_size,
        stride = strides,
        activation_fn = tf.nn.relu,
        padding = padding)
    

    pool = layers.max_pool2d(
        inputs = conv,
        kernel_size = max_pool_window, 
        stride = max_pool_stride)

    return conv, pool


def bottle_neck(
    input_tensor,
    no_filters,
    kernel_size=(3, 3),
    strides=(1, 1),
    padding="SAME"):
    
    
    conv = layers.conv2d(
        inputs = input_tensor,
        num_outputs = no_filters,
        normalizer_fn = layers.batch_norm,
        normalizer_params = {"scale" : True , "fused" : True},
        kernel_size = kernel_size,
        stride = strides,
        activation_fn = tf.nn.relu,
        padding = padding)

    conv = layers.conv2d(
        inputs = conv,
        num_outputs = no_filters,
        normalizer_fn = layers.batch_norm,
        normalizer_params = {"scale" : True , "fused" : True},
        kernel_size = kernel_size,
        stride = strides,
        activation_fn = tf.nn.relu,
        padding = padding)

    return conv




def up_block(
    input_tensor,
    no_filters,
    skip_connection,
    kernel_size=(3, 3),
    strides=(1, 1),
    padding="SAME",
    max_pool_window=(2, 2),
    max_pool_stride=(2, 2)):
    
    
    
    conv = layers.conv2d_transpose(
        inputs = input_tensor,
        num_outputs = no_filters, 
        kernel_size = max_pool_window,
        stride = max_pool_stride,
        padding = padding,
        activation_fn = tf.nn.relu,
        normalizer_fn = layers.batch_norm,
        normalizer_params = {"scale" : True , "fused" : True}
        )
    
    conv = tf.concat([skip_connection, conv], -1)
    
    
    
    conv = layers.conv2d(
        inputs = conv,
        num_outputs = no_filters,
        normalizer_fn = layers.batch_norm,
        normalizer_params = {"scale" : True , "fused" : True},
        kernel_size = kernel_size,
        stride = strides,
        activation_fn = tf.nn.relu,
        padding = padding)

    conv = layers.conv2d(
        inputs = conv,
        num_outputs = no_filters,
        normalizer_fn = layers.batch_norm,
        normalizer_params = {"scale" : True , "fused" : True},
        kernel_size = kernel_size,
        stride = strides,
        activation_fn = tf.nn.relu,
        padding = padding)

    return conv



def output_block(
    input_tensor,
    kernel_size=(3, 3),
    strides=(1, 1),
    padding="SAME"):
    
    
    conv = layers.conv2d(
        inputs = input_tensor,
        num_outputs = 2,
        normalizer_fn = None,
        normalizer_params = None,
        kernel_size = kernel_size,
        stride = strides,
        activation_fn = tf.nn.relu,
        padding = padding)

    conv = layers.conv2d(
        inputs = conv,
        num_outputs = 1,
        normalizer_fn = None,
        normalizer_params = None,
        kernel_size = (1,1),
        stride = strides,
        activation_fn = tf.nn.sigmoid,
        padding = padding)

    return conv 
```
Implementation including ModelSizeRegularizer
```python
def train_model(learning_rate = 0.0001, num_epochs = 100 , minibatch_size = 16):
    
    with tf.Graph().as_default():
        
        
        train_gen = DataGen(path_input = "/content/train2014" , path_mask = "/content/mask_train_2014/" , batch_size = minibatch_size)
        
        val_gen = DataGen(path_input =  "/content/val2014", path_mask =  "/content/mask_val_2014", batch_size = minibatch_size)
        
        train_steps =  int(len(os.listdir( "/content/train2014"))/minibatch_size)
        
        
        X = tf.placeholder(tf.float32, [None, 128, 128, 3])
        Y = tf.placeholder(tf.float32, [None, 128, 128, 1])
                

        filter_size = [64,128,256,512,1024]

        d1 , p1 = down_block(input_tensor= X,
                             no_filters=filter_size[0],
                             kernel_size = (3,3),
                             strides=(1,1),
                             padding="same",
                             max_pool_window=(2,2),
                             max_pool_stride=(2,2))


        d2 , p2 = down_block(input_tensor= p1,
                             no_filters=filter_size[1],
                             kernel_size = (3,3),
                             strides=(1,1),
                             padding="same",
                             max_pool_window=(2,2),
                             max_pool_stride=(2,2))



        d3 , p3 = down_block(input_tensor= p2,
                             no_filters=filter_size[2],
                             kernel_size = (3,3),
                             strides=(1,1),
                             padding="same",
                             max_pool_window=(2,2),
                             max_pool_stride=(2,2))



        d4 , p4 = down_block(input_tensor= p3,
                             no_filters=filter_size[3],
                             kernel_size = (3,3),
                             strides=(1,1),
                             padding="same",
                             max_pool_window=(2,2),
                             max_pool_stride=(2,2))


        b = bottle_neck(input_tensor= p4,
                             no_filters=filter_size[4],
                             kernel_size = (3,3),
                             strides=(1,1),
                             padding="same")



        u4 = up_block(input_tensor = b,
                      no_filters = filter_size[3],
                      skip_connection = d4,
                      kernel_size=(3, 3),
                      strides=(1, 1),
                      max_pool_window = (2,2),
                      max_pool_stride = (2,2),
                      padding="same")

        u3 = up_block(input_tensor = u4,
                      no_filters = filter_size[2],
                      skip_connection = d3,
                      kernel_size=(3, 3),
                      strides=(1, 1),
                      max_pool_window = (2,2),
                      max_pool_stride = (2,2),
                      padding="same")


        u2 = up_block(input_tensor = u3,
                      no_filters = filter_size[1],
                      skip_connection = d2,
                      kernel_size=(3, 3),
                      strides=(1, 1),
                      max_pool_window = (2,2),
                      max_pool_stride = (2,2),
                      padding="same")


        u1 = up_block(input_tensor = u2,
                      no_filters = filter_size[0],
                      skip_connection = d1,
                      kernel_size=(3, 3),
                      strides=(1, 1),
                      max_pool_window = (2,2),
                      max_pool_stride = (2,2),
                      padding="same")



        logits = output_block(input_tensor=u1 ,
                              kernel_size = (3,3),
                              padding = "same",
                              strides = (1,1))

        
        
        network_reg = model_size_regularizer.GammaModelSizeRegularizer([logits.op], gamma_threshold = 1e-3)
        
        regularization_strength = 1e-14
        
        loss = tf.reduce_mean(tf.keras.losses.binary_crossentropy(y_pred=logits , y_true=Y)) + (network_reg.get_regularization_term() * regularization_strength)

        train_op = tf.train.AdamOptimizer(learning_rate = learning_rate).minimize(loss)

        acc = tf.metrics.accuracy(labels = Y , predictions = logits)

        init = tf.global_variables_initializer()

        
        exporter = structure_exporter.StructureExporter(network_reg.op_regularizer_manager)
        
        
        with tf.Session() as sess:
            
            sess.run(init)
            
            counter = 1
            
            for epoch in range(num_epochs):
                
                for batch_no in range(train_steps):
                    
                    x_train , y_train = train_gen.__getitem__(batch_no)
                    
                    _ , structure_exporter_tensors , train_loss   = sess.run([train_op , exporter.tensors , loss ] , feed_dict = {X : x_train , Y : y_train})
                                    
                      
                      
                    print("B : " , batch_no , " Training Loss : " , train_loss )
                    
                    counter = counter + 1
                    
                    if counter % 50 == 0:
                    
                      print("Saving")
                      exporter.populate_tensor_values(structure_exporter_tensors)
                      exporter.create_file_and_save_alive_counts("/content/morph-net/dont_know", counter)


```

But somehow the network size always remains the same even after 5000 iterations..

**Comment by yairmov** [comment]
Can you try setting the regularization strength to be much much stronger. Try 1e-2 and check if the size starts changing then (I expect 1e-2 would be stronger than what you will need eventually, but  for debugging it is useful to use an extreme number).

**Comment by ayp-google** [comment]
Another sanity check is to output the MorphNet loss (e.g., via TensorBoard) and verify that it decreases during training.

**Comment by H-arshit** [comment]
Morphnet is up and working.. Thanks a lot..
